pkg/runtime/wfengine/backend.go

```diff
@@ -0,0 +1,208 @@
+	ScheduleActivity(wi *backend.ActivityWorkItem)
+}
+
+type actorBackend struct {
+	actors                    actors.Actors
+	orchestrationWorkItemChan chan *backend.OrchestrationWorkItem
+	activityWorkItemChan      chan *backend.ActivityWorkItem
+}
+
+func NewActorBackend() *actorBackend {
+	return &actorBackend{
+		orchestrationWorkItemChan: make(chan *backend.OrchestrationWorkItem),
+		activityWorkItemChan:      make(chan *backend.ActivityWorkItem),
+	}
+}
+
+func (be *actorBackend) SetActorRuntime(actors actors.Actors) {
+	be.actors = actors
+}
+
+// ScheduleActivity implements workflowScheduler
+func (be *actorBackend) ScheduleActivity(wi *backend.ActivityWorkItem) {
+	be.activityWorkItemChan <- wi
+}
+
+// ScheduleWorkflow implements workflowScheduler
+func (be *actorBackend) ScheduleWorkflow(wi *backend.OrchestrationWorkItem) {
+	be.orchestrationWorkItemChan <- wi
+}
+
+// CreateOrchestrationInstance implements backend.Backend and creates a new workflow instance.
+//
+// Internally, creating a workflow instance also creates a new actor with the same ID. The create
+// request is saved into the actor's "inbox" and then executed via a reminder thread. If the app is
+// scaled out across multiple replicas, the actor might get assigned to a replicas other than this one.
+func (be *actorBackend) CreateOrchestrationInstance(ctx context.Context, e *backend.HistoryEvent) error {
+	if err := be.validateConfiguration(); err != nil {
+		return err
+	}
+
+	workflowInstanceID := e.GetExecutionStarted().GetOrchestrationInstance().GetInstanceId()
+	if workflowInstanceID == "" {
+		return errors.New("history event must be an ExecutionStartedEvent with a valid orchestration instance ID")
+	}
+
+	eventData, err := backend.MarshalHistoryEvent(e)
+	if err != nil {
+		return err
+	}
+
+	// Invoke the well-known workflow actor directly, which will be created by this invocation
+	// request. Note that this request goes directly to the actor runtime, bypassing the API layer.
+	req := invokev1.
+		NewInvokeMethodRequest(CreateWorkflowInstanceMethod).
+		WithActor(WorkflowActorType, workflowInstanceID).
+		WithRawData(eventData, invokev1.OctetStreamContentType)
+	if _, err := be.actors.Call(ctx, req); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetOrchestrationMetadata implements backend.Backend
+func (be *actorBackend) GetOrchestrationMetadata(ctx context.Context, id api.InstanceID) (*api.OrchestrationMetadata, error) {
+	// Invoke the corresponding actor, which internally stores its own workflow metadata
+	req := invokev1.
+		NewInvokeMethodRequest(GetWorkflowMetadataMethod).
+		WithActor(WorkflowActorType, string(id)).
+		WithRawData(nil, invokev1.OctetStreamContentType)
+	if res, err := be.actors.Call(ctx, req); err != nil {
+		return nil, err
+	} else {
+		_, data := res.RawData()
+		if len(data) == 0 {
+			return nil, api.ErrInstanceNotFound
+		}
+
+		// Decode the data using encoding/gob (https://go.dev/blog/gob)
+		buffer := bytes.NewBuffer(data)
```

Comment:
Using gob is faster than JSON, particularly during decoding, but makes it much harder to debug things where you may need to look at the data; is this a trade-off we want to make? It also means that the only tools that can decode the data have to be written in Go (I think) -- if we care about speed and portability then msgpack may be better if we don't want to use JSON

Reply:
That seems fine, as long as we aren't planning to store it or humans need to interact with it :+1: